Runtime/Tracking/Collision/CollisionTracker.cs

```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
                 return;
             }
 
+            RemoveTrackedNode(collider);
             OnCollisionStopped(eventData.Set(this, true, null, collider));
         }
+
+        /// <summary>
+        /// Adds a <see cref="CollisionTrackerDisabledObserver"/> to the <see cref="GameObject"/> that contains the <see cref="Collider"/> causing the collision.
+        /// </summary>
+        /// <param name="target">The target to add the <see cref="CollisionTrackerDisabledObserver"/> component to.</param>
+        protected virtual void AddTrackedNode(Collider target)
+        {
+            if (target == null || !StopCollisionsOnDisable)
+            {
+                return;
+            }
+
+            trackedCollisions.Add(target);
+            CollisionTrackerDisabledObserver node = target.gameObject.AddComponent<CollisionTrackerDisabledObserver>();
+            node.Source = this;
+            node.Target = target;
+        }
+
+        /// <summary>
+        /// Removes the <see cref="CollisionTrackerDisabledObserver"/> from the <see cref="GameObject"/> that contains the <see cref="Collider"/> ceasing the collision.
+        /// </summary>
+        /// <param name="target">The target to remove the <see cref="CollisionTrackerDisabledObserver"/> component from.</param>
+        protected virtual void RemoveTrackedNode(Collider target)
+        {
+            if (target == null || !StopCollisionsOnDisable)
+            {
+                return;
+            }
+
+            foreach (CollisionTrackerDisabledObserver node in target.gameObject.GetComponents<CollisionTrackerDisabledObserver>().ToArray())
+            {
+                if (node.Source == this && node.Target == target)
+                {
+                    node.Destroy();
+                    trackedCollisions.Remove(target);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Observes the disabled state of any <see cref="GameObject"/> that the <see cref="CollisionTracker"/> is currently colliding with.
+    /// </summary>
+    public class CollisionTrackerDisabledObserver : MonoBehaviour
+    {
+        /// <summary>
+        /// The <see cref="CollisionTracker"/> that is causing the collision.
+        /// </summary>
+        [Serialized, Cleared]
+        [field: DocumentedByXml]
+        public CollisionTracker Source { get; set; }
+        /// <summary>
+        /// The <see cref="Collider"/> that is being collided with.
+        /// </summary>
+        [Serialized, Cleared]
+        [field: DocumentedByXml]
+        public Collider Target { get; set; }
+
+        /// <summary>
+        /// Whether <see cref="this"/> is being destroyed.
+        /// </summary>
+        protected bool isDestroyed;
+
+        /// <summary>
+        /// Destroys <see cref="this"/> from the scene.
+        /// </summary>
+        public virtual void Destroy()

```

Comment:
Shouldn't this instead be `OnDestroyed`? What happens if a user destroys this component, even if it's because of destruction of a parent game object? This logic should still run, shouldn't it?

I may be missing some details around how this magic method works, I know you brought it up on Slack...

Reply:
This isn't the Unity magic method which is `OnDestroy`

This is a custom method that destroys the observer component.

It's so the `CollisionTracker` can remove the component when the collision ends.

The reason it has the `isDestroyed` flag is so I know the component has been removed manually and therefore I don't want the `Source.StopCollision()` method firing when the component is disabled.

Because destroying the object will call `OnDisable` and then you'd get two stop messages..

If you destroy the object that this observer component is on, then `Destroy` isn't called and doesn't set `isDestroyed` so you *do* get the `OnDisable` called and therefore it does correctly call the `StopCollision`

Reply:
I looked into this some more and there is indeed no way to say "OnDisable, but only when *not* destroyed". Too bad Unity always throws in some weird specific implementation detail.